src/blockchain/ain/jellyfish/utils/raw-tx-masternode.ts

```diff
@@ -0,0 +1,61 @@
+import BigNumber from 'bignumber.js';
+import { Config } from 'src/config/config';
+import { Masternode } from 'src/integration/masternode/domain/entities/masternode.entity';
+import { RawTxDto } from '../dto/raw-tx.dto';
+import { RawTxBase } from './raw-tx-base';
+import { RawTxUtil } from './raw-tx-util';
+
+export class RawTxMasternode extends RawTxBase {
+  async create(masternode: Masternode): Promise<RawTxDto> {
+    return this.handle(() => this.createTx(masternode));
+  }
+
+  async resign(masternode: Masternode): Promise<RawTxDto> {
+    return this.handle(() => this.resignTx(masternode));
+  }
+
+  private async createTx(masternode: Masternode): Promise<RawTxDto> {
+    const [ownerScript, ownerPubKeyHash] = RawTxUtil.parseAddress(masternode.owner);
+    const [, operatorPubKeyHash] = RawTxUtil.parseOperatorPubKeyHash(masternode.operator);
+
+    const expectedAmount = new BigNumber(
+      Config.masternode.collateral + Config.masternode.creationFee + Config.masternode.fee,
+    );
+    const utxo = await this.utxoProvider.provideExactAmount(masternode.owner, expectedAmount);
+
+    const vins = RawTxUtil.createVins(utxo.prevouts);
+    const vouts = [
+      RawTxUtil.createVoutCreateMasternode(operatorPubKeyHash, masternode.timeLock),
+      RawTxUtil.createVoutReturn(ownerScript, new BigNumber(Config.masternode.collateral)),
+    ];
+
+    const witnesses = [
+      RawTxUtil.createWitness([
+        RawTxUtil.createWitnessScript(operatorPubKeyHash),
+        RawTxUtil.createWitnessScript(ownerPubKeyHash),
+      ]),
+    ];
+
+    return RawTxUtil.generateTx(utxo, vins, vouts, witnesses);
+  }
+
+  private async resignTx(masternode: Masternode): Promise<RawTxDto> {
+    const [, ownerPubKeyHash] = RawTxUtil.parseAddress(masternode.owner);
+    const [, operatorPubKeyHash] = RawTxUtil.parseOperatorPubKeyHash(masternode.operator);
+
+    const expectedAmount = new BigNumber(Config.masternode.resignFee);
+    const utxo = await this.utxoProvider.provideExactAmount(masternode.owner, expectedAmount);
+
+    const vins = RawTxUtil.createVins(utxo.prevouts);
+    const vouts = [RawTxUtil.createVoutResignMasternode(masternode.creationHash)];
+
+    const witnesses = [
+      RawTxUtil.createWitness([
+        RawTxUtil.createWitnessScript(operatorPubKeyHash),
+        RawTxUtil.createWitnessScript(ownerPubKeyHash),
+      ]),
+    ];
+
+    return RawTxUtil.generateTx(utxo, vins, vouts, witnesses);
+  }
```

Comment:
These are the same except the amount and the vouts. Maybe combine them? Maybe for creation use
```ts
async create(masternode: Masternode): Promise<RawTxDto> {
    const amount = new BigNumber(Config.masternode.collateral + Config.masternode.creationFee + Config.masternode.fee);
    const vouts = this.getCreateVouts(masternode);
    return this.handle(() => this.createTx(amount, vouts));
  }
```

Reply:
Partially agree, a few parts are the same. But you would need the `operatorPubKeyHash` and `ownerScript` to create the vouts. Which would later be reused for witnesses. So either you parse them twice or you move out all of them into the `createVouts`. Yes we have a little bit of code duplication right now, but IMHO it makes it more readable.

If we would create a vout functions, which either creates `CreateMasternode` or `ResignMasternode` vouts, it would make it way less readable.

Reply:
Ok agreed